### PR TITLE
[Merged by Bors] - feat: port AlgebraicGeometry.Spec

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

- [x] depends on: #4693

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jcommelin** [comment]
Here the proofs are blowing up as well. I think we should take a step back and debug what the issue is. Otherwise AG in mathlib4 is going to be quite a bit more painful than in mathlib3.

**Comment by jjaassoonn** [comment]
> Here the proofs are blowing up as well. I think we should take a step back and debug what the issue is. Otherwise AG in mathlib4 is going to be quite a bit more painful than in mathlib3.

I wonder if changing `ConcreteCategory`'s coercion of function to `FunLike` would help. I will test it out later

**Comment by jjaassoonn** [comment]
> Here the proofs are blowing up as well. I think we should take a step back and debug what the issue is. Otherwise AG in mathlib4 is going to be quite a bit more painful than in mathlib3.

I changed `ConcreteCategory`'s way of coercing function to `FunLike`, it definitely makes rewriting `comp_apply` easier without making other proof much longer (mostly add `comp_apply`, `id_apply` to some `simp` call). Please tell me what you think @jcommelin

**Comment by jcommelin** [comment]
@jjaassoonn As far as I can tell, this is a very good move! But please make this refactor in a separate PR, so that the current PR is only about porting `Spec`.

**Comment by jjaassoonn** [comment]
> @jjaassoonn As far as I can tell, this is a very good move! But please make this refactor in a separate PR, so that the current PR is only about porting `Spec`.

I have opened this new PR at #4693

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4693
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4693~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/AlgebraicGeometry/Spec.lean [hunk 1]
```diff
@@ -112,100 +121,129 @@ theorem Spec.sheafedSpaceMap_id {R : CommRingCat} :
...
+set_option linter.uppercaseLean3 false in
 #align algebraic_geometry.Spec.SheafedSpace_map_id AlgebraicGeometry.Spec.sheafedSpaceMap_id
 
 theorem Spec.sheafedSpaceMap_comp {R S T : CommRingCat} (f : R ⟶ S) (g : S ⟶ T) :
     Spec.sheafedSpaceMap (f ≫ g) = Spec.sheafedSpaceMap g ≫ Spec.sheafedSpaceMap f :=
   AlgebraicGeometry.PresheafedSpace.Hom.ext _ _ (Spec.topMap_comp f g) <|
-    NatTrans.ext _ _ <|
-      funext fun U => by dsimp; rw [CategoryTheory.Functor.map_id]; rw [category.comp_id];
-        erw [comap_comp f g]; rfl
+    NatTrans.ext _ _ <| funext fun U =>
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/Spec.lean:133]: This deserves a porting note. It went from 3 lines to 13.

File: Mathlib/AlgebraicGeometry/Spec.lean [hunk 2]
```diff
@@ -228,63 +267,84 @@ def Spec.locallyRingedSpaceMap {R S : CommRingCat} (f : R ⟶ S) :
...
-      replace ha := IsLocalRingHom.map_nonunit _ ha
-      convert RingHom.isUnit_map (stalk_iso R (PrimeSpectrum.comap f p)).inv ha
-      rw [iso.hom_inv_id_apply]
+      -- *locally* ringed spaces, i.e. that the induced map on the stalks is a local ring
+      -- homomorphism.
+      erw [← localRingHom_comp_stalkIso_apply] at ha
+      replace ha := (stalkIso S p).hom.isUnit_map ha
+      rw [← comp_apply, show localizationToStalk S p = (stalkIso S p).inv from rfl,
+        Iso.inv_hom_id, id_apply] at ha
+      replace ha := IsLocalRingHom.map_nonunit
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/Spec.lean:276]:
```suggestion
      -- Porting note: `R` and `S` had to be made explicit
      replace ha := IsLocalRingHom.map_nonunit
```

File: Mathlib/AlgebraicGeometry/Spec.lean [hunk 3]
```diff
@@ -228,63 +267,84 @@ def Spec.locallyRingedSpaceMap {R S : CommRingCat} (f : R ⟶ S) :
...
 
 /-- The counit (`Spec_Γ_identity.inv.op`) of the adjunction `Γ ⊣ Spec` is an isomorphism. -/
-@[simps hom_app inv_app]
+@[simps! hom_app inv_app]
 def specΓIdentity : Spec.toLocallyRingedSpace.rightOp ⋙ Γ ≅ 𝟭 _ :=
-  Iso.symm <| NatIso.ofComponents (fun R => asIso (toSpecΓ R) : _) fun _ _ => Spec_Γ_naturality
+  Iso.symm <| NatIso.ofComponents (fun R =>
+    -- Porting note : In Lean3, this `IsIso` is synthesized automatically
+    letI : IsIso (toSpecΓ R) := StructureSheaf.isIso_to_global _
+    asIso (toSpecΓ R)) fun {X Y} f => by
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/Spec.lean:343]:
```suggestion
    -- Porting note: In mathlib 3, the tactic block wasn't needed: `Spec_Γ_naturality` type-checked as term
    asIso (toSpecΓ R)) fun {X Y} f => by
```

**Comment by jcommelin** [comment]
I left 3 comments about porting notes. Apart from that, this looks ready to go now.

Thanks for battling through this and for doing the refactor that made the diff much nicer!

File: Mathlib/AlgebraicGeometry/Spec.lean (hunk 1 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Spec.lean:133]: I have made the proof shorter, this is an example where `dsimp` behaves very differently from its `Lean3` behaviour

File: Mathlib/AlgebraicGeometry/Spec.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Spec.lean:276]: After second check, only `R` needs to be made explicit

File: Mathlib/AlgebraicGeometry/Spec.lean (hunk 3 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Spec.lean:343]: This behaviour is slightly confusing, `by exact Spec_Γ_naturality (R := X) (S := Y) f` works but `Spec_Γ_naturality (R := X) (S := Y) f` does not work

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**